### PR TITLE
implemented ShoppingCart and CartItem entities and services

### PR DESCRIPTION
**Pull request by JeniaSan**

### CODE REVIEW

File: src/main/java/com/polishuchenko/bookstore/service/cartitem/impl/CartItemServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    public CartItemResponseDto update(Long id, UpdateCartItemDto updateCartItemDto) {
+        CartItem cartItem = cartItemRepository.findCartItemById(id).orElseThrow(
```
Comment (vovasalyha) [inline src/main/java/com/polishuchenko/bookstore/service/cartitem/impl/CartItemServiceImpl.java]: What if user will specify ID of the item of other user?

File: src/main/java/com/polishuchenko/bookstore/service/cartitem/impl/CartItemServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import com.polishuchenko.bookstore.mapper.CartItemMapper;
+import com.polishuchenko.bookstore.model.CartItem;
+import com.polishuchenko.bookstore.repository.cartitem.CartItemRepository;
+import com.polishuchenko.bookstore.service.cartitem.CartItemService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
```
Comment (vovasalyha) [inline src/main/java/com/polishuchenko/bookstore/service/cartitem/impl/CartItemServiceImpl.java]: I think it's redundant to create separate service. Put the methods to the same service where you put `addItem` method.

File: src/main/java/com/polishuchenko/bookstore/service/user/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -32,6 +36,15 @@ public UserRegistrationResponseDto register(UserRegistrationRequestDto request)
         user.setPassword(passwordEncoder.encode(request.password()));
         Role role = roleRepository.getByName(Role.RoleName.ROLE_USER);
         user.setRoles(Set.of(role));
-        return userMapper.toResponseUser(userRepository.save(user));
+        User savedUser = userRepository.save(user);
+        ShoppingCart shoppingCart = new ShoppingCart();
```
Comment (vovasalyha) [inline src/main/java/com/polishuchenko/bookstore/service/user/impl/UserServiceImpl.java]: I suggest to extract shopping cart related logic to a private method

File: src/main/resources/db/changelog/changes/create-shopping_cart-table.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    nullable: false
+                    primaryKey: true
+              - column:
+                  name: user_id
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/create-shopping_cart-table.yaml:17]: where is foreign key to the user table?

**Review by vovasalyha — CHANGES_REQUESTED** [review]
GJ. Let's improve it a bit

File: src/main/resources/db/changelog/changes/create-shopping_cart-table.yaml (hunk 4 above)
Comment (JeniaSan) [reply src/main/resources/db/changelog/changes/create-shopping_cart-table.yaml:17]: https://mateacademy.slack.com/archives/C05B8U42C69/p1694610657937319

File: src/main/java/com/polishuchenko/bookstore/dto/cartitem/UpdateCartItemDto.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.polishuchenko.bookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class UpdateCartItemDto {
```
Comment (andrii-hoienko) [inline src/main/java/com/polishuchenko/bookstore/dto/cartitem/UpdateCartItemDto.java:8]:
```suggestion
public class UpdateRequestCartItemDto {
```

File: src/main/java/com/polishuchenko/bookstore/mapper/ShoppingCartMapper.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    @AfterMapping
+    default void cartItemsToDto(
+            @MappingTarget ShoppingCartDto cartDto, ShoppingCart shoppingCart) {
+        Set<CartItemResponseDto> itemsDtos = shoppingCart.getCartItems().stream()
+                .map(this::mapToDto)
+                .collect(Collectors.toSet());
+        cartDto.setCartItems(itemsDtos);
+    }
+
+    default CartItemResponseDto mapToDto(CartItem cartItem) {
```
Comment (andrii-hoienko) [inline src/main/java/com/polishuchenko/bookstore/mapper/ShoppingCartMapper.java]: You have a specific class for CartItem mapping is it possible to move all CartItem mapping logic there?

File: src/main/java/com/polishuchenko/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final UserService userService;
+    private final CartItemMapper cartItemMapper;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    public CartItemResponseDto update(Long id, UpdateCartItemDto updateCartItemDto) {
```
Comment (andrii-hoienko) [inline src/main/java/com/polishuchenko/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:36]: Think carefully witch methods should be transactional

File: src/main/java/com/polishuchenko/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart() {
+        return shoppingCartMapper.toDto(currentUserCart());
+    }
+
+    @Transactional
+    @Override
+    public CartItemResponseDto addCartItem(CartItemRequestDto cartItemRequestDto) {
```
Comment (andrii-hoienko) [inline src/main/java/com/polishuchenko/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:67]: To many responsibilities for this method, I suggest moving logic or increasing quantity of cartItem in separate method

File: src/main/java/com/polishuchenko/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        } else {
+            cartItem = new CartItem();
+            cartItem.setShoppingCart(shoppingCart);
+            cartItem.setBook(book);
+        }
+        cartItem.setQuantity(cartItemRequestDto.quantity());
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    private ShoppingCart currentUserCart() {
```
Comment (andrii-hoienko) [inline src/main/java/com/polishuchenko/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
    private ShoppingCart getCurrentUserCart() {
```

File: src/main/java/com/polishuchenko/bookstore/service/user/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -32,6 +36,19 @@ public UserRegistrationResponseDto register(UserRegistrationRequestDto request)
...
+        createShoppingCart(savedUser);
+        return userMapper.toResponseUser(savedUser);
+    }
+
+    @Override
+    public User getAuthenticatedUser() {
+        return (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+    }
+
+    private void createShoppingCart(User user) {
```
Comment (andrii-hoienko) [inline src/main/java/com/polishuchenko/bookstore/service/user/impl/UserServiceImpl.java]: user service is responsible for shopping cart creation?

File: src/main/resources/db/changelog/changes/create-cart_item-table.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-cart_item-table
+      author: YEvhenii
```
Comment (andrii-hoienko) [inline src/main/resources/db/changelog/changes/create-cart_item-table.yaml]: YEvhenii Henii, but please specify your surname as well

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, but some improvements required

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**